v3/src/components/graph/components/chartdots.tsx

```diff
@@ -1,126 +1,157 @@
 }
 
+type BinMap = Record<string, Record<string, number>>
+
 export const ChartDots = memo(function ChartDots(props: IProps) {
   const {graphModel, plotProps: {dotsRef, enableAnimation}} = props,
     dataConfig = useDataConfigurationContext(),
     dataset = useDataSetContext(),
     layout = useGraphLayoutContext(),
-    xAttrID = dataConfig?.attributeID('x'),
-    xAttributeType = dataConfig?.attributeType('x'),
-    yAttrID = dataConfig?.attributeID('y'),
-    // yAttributeType = dataset?.attrFromID(yAttrID)?.type,
-    primaryPlace = xAttributeType === 'categorical' ? 'bottom' : 'left',
-    primaryAttributeID = primaryPlace === 'left' ? yAttrID : xAttrID,
-    secondaryPlace = primaryPlace === 'left' ? 'bottom' : 'left',
-    primaryScale = layout.axisScale(primaryPlace) as ScaleBand<string>,
-    secondaryScale = layout.axisScale(secondaryPlace) as ScaleType,
-    attribute = primaryAttributeID && dataset?.attrFromID(primaryAttributeID) || undefined,
-    categories = Array.from(new Set(attribute?.strValues)),
-    categoriesMapRef = useRef<Record<string, { cell: number, numSoFar: number }>>({})
+    primaryAttrPlace = dataConfig?.primaryPlace ?? 'x',
+    primaryAxisPlace = attrPlaceToAxisPlace[primaryAttrPlace] ?? 'bottom',
+    primaryIsBottom = primaryAxisPlace === 'bottom',
+    primaryAttrID = dataConfig?.attributeID(primaryAttrPlace),
+    secondaryAttrPlace = primaryAttrPlace === 'x' ? 'y' : 'x',
+    secondaryAxisPlace = attrPlaceToAxisPlace[secondaryAttrPlace] ?? 'left',
+    secondaryAttrID = dataConfig?.attributeID(secondaryAttrPlace),
+    primaryScale = layout.axisScale(primaryAxisPlace) as ScaleBand<string>,
+    secondaryScale = layout.axisScale(secondaryAxisPlace) as ScaleBand<string>
 
   const computeMaxOverAllCells = useCallback(() => {
-    const values = attribute?.strValues,
-      bins: Record<string, number> = {}
-    values?.forEach(aValue => {
-      if (bins[aValue] === undefined) {
-        bins[aValue] = 0
+    const valuePairs = dataConfig?.cases.map(caseID => {
+        return {
+          primary: (primaryAttrID && dataset?.getValue(caseID, primaryAttrID)) ?? '',
+          secondary: (secondaryAttrID && dataset?.getValue(caseID, secondaryAttrID)) ?? '__main__'
+        }
+      }),
+      bins: BinMap = {}
+    valuePairs?.forEach(aValue => {
+      if (bins[aValue.primary] === undefined) {
+        bins[aValue.primary] = {}
+      }
+      if (bins[aValue.primary][aValue.secondary] === undefined) {
+        bins[aValue.primary][aValue.secondary] = 0
       }
-      bins[aValue]++
+      bins[aValue.primary][aValue.secondary]++
     })
-    return Object.keys(bins).reduce((max, aKey) => Math.max(max, bins[aKey]), 0)
-  }, [attribute])
+    return Object.keys(bins).reduce((hMax, hKey) => {
+      return Math.max(hMax, Object.keys(bins[hKey]).reduce((vMax, vKey) => {
+        return Math.max(vMax, bins[hKey][vKey])
+      }, 0))
+    }, 0)
+  }, [dataConfig?.cases, dataset, primaryAttrID, secondaryAttrID])
```

Comment:
Use the `const { cases } = dataConfig` trick above so that this just depends on `dataConfig`.